src/Middleware/HttpLogging/src/ResponseBufferingStream.cs

```diff
@@ -0,0 +1,446 @@
+                _currentSegment = null;
+            }
+
+            _bytesWritten = 0;
+            _position = 0;
+        }
+
+        public void Advance(int count)
+        {
+            _bytesWritten += count;
+            _position += count;
+        }
+
+        public Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            EnsureCapacity(sizeHint);
+
+            return _currentSegment.AsMemory(_position, _currentSegment.Length - _position);
+        }
+
+        public Span<byte> GetSpan(int sizeHint = 0)
+        {
+            EnsureCapacity(sizeHint);
+
+            return _currentSegment.AsSpan(_position, _currentSegment.Length - _position);
+        }
+
+        public void CopyTo(IBufferWriter<byte> destination)
+        {
+            if (_completedSegments != null)
+            {
+                // Copy completed segments
+                var count = _completedSegments.Count;
+                for (var i = 0; i < count; i++)
+                {
+                    destination.Write(_completedSegments[i].Span);
+                }
+            }
+
+            destination.Write(_currentSegment.AsSpan(0, _position));
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (_completedSegments == null && _currentSegment is not null)
+            {
+                // There is only one segment so write without awaiting.
+                return destination.WriteAsync(_currentSegment, 0, _position);
+            }
+
+            return CopyToSlowAsync(destination);
+        }
+
+        [MemberNotNull(nameof(_currentSegment))]
+        private void EnsureCapacity(int sizeHint)
+        {
+            // This does the Right Thing. It only subtracts _position from the current segment length if it's non-null.
+            // If _currentSegment is null, it returns 0.
+            var remainingSize = _currentSegment?.Length - _position ?? 0;
+
+            // If the sizeHint is 0, any capacity will do
+            // Otherwise, the buffer must have enough space for the entire size hint, or we need to add a segment.
+            if ((sizeHint == 0 && remainingSize > 0) || (sizeHint > 0 && remainingSize >= sizeHint))
+            {
+                // We have capacity in the current segment
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
+                return;
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
+            }
+
+            AddSegment(sizeHint);
+        }
+
+        [MemberNotNull(nameof(_currentSegment))]
+        private void AddSegment(int sizeHint = 0)
+        {
+            if (_currentSegment != null)
+            {
+                // We're adding a segment to the list
+                if (_completedSegments == null)
+                {
+                    _completedSegments = new List<CompletedBuffer>();
+                }
```

Comment:
I don't know if I'd have a default better. Maybe limit / 4096

Reply:
❔ I'd do some debug-logging for the count and ran some representative workload over it, to see what value seems proper.
Right now it's just guessing.

The concern I have is that the default capacity of 4 completed-segments won't fit, so a new backing-array for the list of size 8 is allocated, copied and the previous with size 4 is GCed. And so on with 16, 32, ...

If 4 is reasonable choice, everything is good. Otherwise we could save some allocations in the common case by setting the capacity.

Reply:
> Maybe limit / 4096

Now I got it (it's a division 😉). Sounds good as upper-bound, but then it should be pooled.
Otherwise we might pay too big allocation if not needed?